Add default args to Signed constructors

jku opened the pull request:
Tweak the constructor signatures to make creating brand new Metadata easier (the constructors were originally designed for deserialization purposes). Modify examples and tests to use the new constructor args.

The changes are backwards compatible. Resulting API is much easier to use for creating new Metadata. The API is still a bit unusual but not difficult to learn IMO:

    targets = Metadata(Targets())

There are some noteworthy cases:
* `expires` does not have a reasonable default value and I planned to leave it a required argument. That would have broken backwards compat however as order of arguments would have had to change. So expires now defaults to an arbitrary 1 day in future _EDIT: changed to `utcnow()` so default is expired metadata_
* `consistent_snapshot` defaults to True but None is still possible to indicate Metadata does not contain the field at all
* `spec_version` defaults to whatever is the current library supported version

Only the core metadata classes we're updated. There are other classes that would probably benefit from a better constructor, like Delegations, but those were not handled yet. 

Fixes #1459

**Please verify and check that the pull request fulfills the following
requirements**:

- [x] The code follows the [Code Style Guidelines](https://github.com/secure-systems-lab/code-style-guidelines#code-style-guidelines)
- [ ] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

Commit:
Metadata API: Add default constructor arguments

This allows creating new metadata with less boilerplate:
    root = Metadata(Root())
    targets = Metadata(Targets())

Set reasonable default values for all the arguments -- version to
1, spec_version to current supported version, etc.

Expires does not have a good default value and my original plan was
to require expires argument to be set. That would mean an
incompatible API change though as arguments before expires would be
now optional... So expires now defaults to an arbitrary value of 1
day from moment of creation.

One noteworthy special case is consistent_snapshot where the default
value is True (since that's what we want people to use for new
metadata) but None is also used to imply that metadata does not contain
the field at all.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
tests: Use the default Metadata constructor args

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
examples: Use the constructor default arguments

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
> * [ ]  Tests have been added for the bug fix or new feature

The generated tests and serialization tests do a good job testing this I believe.

> * [x]  Docs have been added for the bug fix or new feature

Metadata docstring should maybe be improved to show an example of creating a new Metadata...

jku commented on tuf/api/metadata.py:
```diff
@@ -120,11 +120,11 @@ class Metadata(Generic[T]):
     ):
         self.signed: T = signed
-        self.signatures = signatures
+        self.signatures = signatures or {}
```
this construct is a bit problematic in a specific edge case of the argument container being empty:

```python
sigs = {}
root = Metadata(Root(), sigs)
assert sigs is root.signatures
```

so maybe I should check `is None` instead

MVrachev replied:
Yea, I think this:
`self.signatures = signatures if signatures is not None else {}`
would be better.

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -497,11 +499,15 @@ def __init__(
-        self.expires = expires
 
-        if version <= 0:
+        self.expires = expires or datetime.utcnow() + timedelta(days=1)
```
My personal preference is if we use a default value of one year for example.
I am a little worried for an instance when somebody forgets passing `expires` and `expires` becomes only one day...

MVrachev replied:
Also, I think it's worth noting in the docstring that if no expires is given a default value of `X` amount of days/years will be used.
Or maybe we don't want to promote that expires is optional and not mention it?

jku replied:
There is no good default value: it's not clear at all that a years expiry is better than one days expiry -- at least a short one won't be used in production for a year before the mistake is noticed. The safe thing would be to set expiry to `utcnow()` so metadata is immediately expired, that's certainly an option... Maybe an option I like the most.

The decision is not very important I think? More important is that we show what should be done in examples and docs.

MVrachev replied:
If we do want to maintain the compatibility of this feature, yes, it makes sense to make it expire as soon as possible.

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -497,11 +499,15 @@ def __init__(
+        self.expires = expires or datetime.utcnow() + timedelta(days=1)
+
+        if version is None:
+            version = 1
```
The same question about the docstring and default values applies to all attributes.
Do we want to mention the default values?

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -1660,15 +1668,15 @@ class Targets(Signed):
     ) -> None:
         super().__init__(version, spec_version, expires, unrecognized_fields)
-        self.targets = targets
+        self.targets = targets or {}
```
Do you think it would be a good idea to check for `None` again instead of a truthy value the same way as you suggested about `signatures`?
Maybe a user could want to just pass `{}` targets?

jku replied:
yeah all these constructs are a bit suspect when the field is a container, checking for None is probably best

MVrachev commented on tests/generated_data/generate_md.py:
```diff
@@ -97,23 +96,15 @@ def generate_all_files(
-
-    targets = Targets(1, SPEC_VERSION, EXPIRY, {})
-    md_targets: Metadata[Targets] = Metadata(targets, {})
+    md_root = Metadata(Root(expires=EXPIRY))
```
I think because of the use of default values for Metadata and Root attributes you can remove `ROLES` as well (line 69).

jku replied:
right, same is likely true for SPEC_VERSION. 

Lint won't complain about unused variables because they're globals so they get exported as module variables. That's a bit annoying, I wonder if there's a best practice here...

MVrachev commented on tests/generated_data/generate_md.py:
```diff
@@ -97,23 +96,15 @@ def generate_all_files(
-    targets = Targets(1, SPEC_VERSION, EXPIRY, {})
-    md_targets: Metadata[Targets] = Metadata(targets, {})
+    md_root = Metadata(Root(expires=EXPIRY))
+    md_timestamp = Metadata(Timestamp(expires=EXPIRY))
+    md_snapshot = Metadata(Snapshot(expires=EXPIRY))
+    md_targets = Metadata(Targets(expires=EXPIRY))
```
I really like you don't have to create temporary `Signed` objects but you can't directly instantiate a Metadata object with populated default `Signed` fields as well. 
Really cool!

MVrachev reviewed (commented):
Overall, the changes are great with a few comments.

I am wondering whether we should simplify and use the default attribute values in the `examples` scripts?
For those examples, I think we all agree the most important priority is that the code is easy to understand.
Otherwise, the simplifications are great, no doubt.

Commit:
tests: Remove unused variables from generate_md

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: Be more careful with container args

If argument is an empty container, we want to use the given empty
container. Only create a new container if argument is None.

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: Set default expires to utcnow()

This means the metadata is by default expired: this seems like a fine
default since we only allow a default value for practical reasons (not
allowing it would mean backwards incompatible API change).

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

Commit:
Metadata API: Document constructor default arguments

Signed-off-by: Jussi Kukkonen <jkukkonen@vmware.com>

jku commented:
I believe all comments we're resolved with additional commits:
* more docs for the default values
* more care taken with empty containers as args (this issue is theoretically there for unrecognised_fields as well btw)
* expires default is `utcnow()` (so expired out of the gate)
* some unused variables cleaned up

jku commented:
> I am wondering whether we should simplify and use the default attribute values in the `examples` scripts?

I believe this is done, if you have specific place you're thinking of let me know.

MVrachev commented on tuf/api/metadata.py:
```diff
@@ -1183,12 +1199,14 @@ class Snapshot(Signed):
         unrecognized_fields: Dictionary of all attributes that are not managed
             by TUF Metadata API
         meta: Dictionary of target metadata filenames to ``MetaFile`` objects.
+            Default is a dictionary with a Metafile for "snapshot.json".
```
```suggestion
            Default is a dictionary with a Metafile for "snapshot.json" with version 1.
```

MVrachev commented:
> > I am wondering whether we should simplify and use the default attribute values in the `examples` scripts?
> 
> I believe this is done, if you have specific place you're thinking of let me know.

I am sorry, I think I was unclear.
I meant do we want to change anything inside the `examples` folders at all?
My point was that the primary purpose of all examples scripts is to be easily understandable and this new syntax using the default values could hide some details.

jku commented:
> examples scripts is to be easily understandable and this new syntax using the default values could hide some details.

If there is a specific case where this is true I'm willing  to discuss. I just don't see one, everything seems to have a fairly obvious default value apart from expiry.

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -104,6 +104,14 @@ class Metadata(Generic[T]):
     ``[Root]`` is not validated at runtime (as pure annotations are not available
     then).
 
+    New Metadata instances can created from scratch with::
```
```suggestion
    New Metadata instances can be created from scratch with::
```

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -444,9 +453,10 @@ class Signed(metaclass=abc.ABCMeta):
+        version: Metadata version number. Default is 1.
+        spec_version: Supported TUF specification version. Default is the
+            version currently supported by the library.
+        expires: Metadata expiry date. Default is current date and time.
```
Is it worth mention that these are optional in the child classes only? I was confused for a split-second, because of the `Optional` type hint but no default value in the constructor signature (I understand now that [`Optional[X]` just means `X | None`](https://docs.python.org/3/library/typing.html#typing.Optional) regardless of it being an optional arg). What about something like...

```suggestion
        expires: Metadata expiry date. If None is passed the current date and time is assigned.
```
(same for `version` and `spec_version`)

jku replied:
Maybe that would be better -- the small but important difference between Optional and the argument actually being optional is indeed annoying and hard to keep clear

lukpueh reviewed (changes requested):
I love the change! Please fix typo (see inline) and this is good to go. 

The comment about the docstring is not a blocker nor is this one:
Re beeing more careful with empty containers (d8c0f3b3f39d11a406883164b0d32d4346ba711f), should we also update all the `unrecognized_fields or {}` in Metadata constructors?

jku commented:
Force pushed with the documentation commit amended with three fixes for issues pointed out in comments (the last three resolved discussions).

> Re beeing more careful with empty containers (https://github.com/theupdateframework/python-tuf/commit/d8c0f3b3f39d11a406883164b0d32d4346ba711f), should we also update all the unrecognized_fields or {} in Metadata constructors?

I'll file an issue: it's pretty minor but strictly speaking yes we should

lukpueh commented on tuf/api/metadata.py:
```diff
@@ -444,9 +453,11 @@ class Signed(metaclass=abc.ABCMeta):
-        version: Metadata version number.
-        spec_version: Supported TUF specification version number.
-        expires: Metadata expiry date.
+        version: Metadata version number. If None, 1 one is assigned.
```
```suggestion
        version: Metadata version number. If None, 1 is assigned.
```
?

jku replied:
damn it :facepalm:

lukpueh replied:
🤷  happens

jku commented:
Once more, force-pushed with a fix for latest resolved review comment

lukpueh reviewed (approved):
Thanks!
